File: go/font_loader.go
package main

import (
	"encoding/json"
	"fmt"
	"image"
	"image/color"
	_ "image/png"
)

type fontMeta struct {
	Width          int      `json:"width"`
	Height         int      `json:"height"`
	BlankChar      string   `json:"blankChar"`
	PixelColor     string   `json:"pixelColor"`
	CharSpacing    [2]int   `json:"charSpacing"`
	CharacterLines []string `json:"characterLines"`
}

func LoadFont(name string, fontDir string) (*Font, error) {
	// Load JSON metadata
	jsonData, err := resources.ReadFile(fontDir + "/" + name + ".json")
	if err != nil {
		return nil, fmt.Errorf("reading font metadata: %w", err)
	}
	var meta fontMeta
	if err := json.Unmarshal(jsonData, &meta); err != nil {
		return nil, fmt.Errorf("parsing font metadata: %w", err)
	}

	// Load PNG image
	pngFile, err := resources.Open(fontDir + "/" + name + ".png")
	if err != nil {
		return nil, fmt.Errorf("opening font image: %w", err)
	}
	defer pngFile.Close()
	img, _, err := image.Decode(pngFile)
	if err != nil {
		return nil, fmt.Errorf("decoding font image: %w", err)
	}

	// Parse pixel color from hex string like "#333333"
	pixelColor := parseHexColor(meta.PixelColor)

	charBitmaps := make(map[int][]int)
	charPixels := meta.Width * meta.Height
	oy := 0

	for _, charLine := range meta.CharacterLines {
		ox := 0
		for i := 0; i < len(charLine); i++ {
			ch := charLine[i]
			if string(ch) == meta.BlankChar {
				ox += meta.Width + meta.CharSpacing[0]
				continue
			}
			pixels := make([]int, charPixels)
			for y := 0; y < meta.Height; y++ {
				for x := 0; x < meta.Width; x++ {
					r, g, b, _ := img.At(ox+x, oy+y).RGBA()
					pr, pg, pb, _ := pixelColor.RGBA()
					if r == pr && g == pg && b == pb {
						pixels[y*meta.Width+x] = FontFG
					}
				}
			}
			charBitmaps[int(ch)] = pixels
			ox += meta.Width + meta.CharSpacing[0]
		}
		oy += meta.Height + meta.CharSpacing[1]
	}

	return &Font{
		CharBitmaps: charBitmaps,
		W:           meta.Width,
		H:           meta.Height,
		BlankChar:   int(meta.BlankChar[0]),
	}, nil
}

func parseHexColor(hex string) color.RGBA {
	if len(hex) == 7 && hex[0] == '#' {
		hex = hex[1:]
	}
	var r, g, b uint8
	fmt.Sscanf(hex, "%02x%02x%02x", &r, &g, &b)
	return color.RGBA{R: r, G: g, B: b, A: 255}
}

File: php/pixelpong/bitmap/FontLoader.php
<?php

namespace stigsb\pixelpong\bitmap;

class FontLoader
{
    const SKIP_CHAR = 32;
    const NEWLINE_CHAR = 10;

    private $fontDir;

    public function __construct($fontDir)
    {
        $this->fontDir = $fontDir;
    }

    public function loadFont($name)
    {
        $png_file = "{$this->fontDir}/{$name}.png";
        $image = imagecreatefrompng($png_file);
//        $sx = imagesx($image);
//        $sy = imagesy($image);
        $font_meta = $this->getFontMetaData($name);
        $oy = 0;
        $char_pixels = $font_meta->width * $font_meta->height;
        $character_bitmaps = [];
        $pixel_color = $font_meta->pixelColor;
        if (preg_match('/^#([0-9a-f]{6})$/', $pixel_color, $m)) {
            $pixel_color = hexdec($m[1]);
        }
        foreach ($font_meta->characterLines as $charLine) {
            for ($i = 0, $ox = 0; $i < strlen($charLine); ++$i, $ox += $font_meta->width + $font_meta->charSpacing[0]) {
                if ($charLine[$i] == $font_meta->blankChar) {
                    continue;
                }
                $pixels = array_fill(0, $char_pixels, Font::BG);
                for ($y = 0; $y < $font_meta->height; ++$y) {
                    for ($x = 0; $x < $font_meta->width; ++$x) {
                        if (imagecolorat($image, $ox + $x, $oy + $y) == $pixel_color) {
                            $pixels[($y * $font_meta->width) + $x] = Font::FG;
                        }
                    }
                }
                $character_bitmaps[ord($charLine[$i])] = $pixels;
            }
            $oy += $font_meta->height + $font_meta->charSpacing[1];
        }
        return $this->createFontWithMetaDataAndBitmaps($font_meta, $character_bitmaps);
    }

    protected function getFontMetaData($name)
    {
        $charmap_file = "{$this->fontDir}/{$name}.json";
        return json_decode(file_get_contents($charmap_file));
    }

    protected function createFontWithMetaDataAndBitmaps($font_meta, $character_bitmaps)
    {
        return new Font($character_bitmaps, $font_meta->width, $font_meta->height, ord($font_meta->blankChar));
    }

}